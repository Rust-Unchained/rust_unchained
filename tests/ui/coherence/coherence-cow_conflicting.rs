//@ aux-build:coherence_lib.rs

// Test that the `Pair` type reports an error if it contains type
// parameters, even when they are covered by local types. This test
// was originally intended to test the opposite, but the rules changed
// with RFC 1023 and this became illegal.

extern crate coherence_lib as lib;
use lib::{Remote,Pair};

pub struct Cover<T>(T);

// It may seem like it, but the two bellow can never conflict:
impl<T> Remote for Pair<T,Cover<T>> { }

impl<T> Remote for Pair<Cover<T>,T> { }

// This one conflicts though
impl<T,U> Remote for Pair<Cover<T>,U> { }
//~^ ERROR E0119

fn main() { }
