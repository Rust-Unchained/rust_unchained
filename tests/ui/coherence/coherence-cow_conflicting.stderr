error[E0119]: conflicting implementations of trait `Remote` for type `Pair<Cover<_>, Cover<Cover<_>>>`
  --> $DIR/coherence-cow_conflicting.rs:19:1
   |
LL | impl<T> Remote for Pair<T,Cover<T>> { }
   | ----------------------------------- first implementation here
...
LL | impl<T,U> Remote for Pair<Cover<T>,U> { }
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `Pair<Cover<_>, Cover<Cover<_>>>`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0119`.
